main/src/ca/uwaterloo/flix/tools/pkg/ManifestParser.scala

```diff
@@ -16,22 +16,218 @@
 package ca.uwaterloo.flix.tools.pkg
 
 import ca.uwaterloo.flix.util.Result
-import org.tomlj.Toml
+import ca.uwaterloo.flix.util.Result.{Err, Ok}
+import org.tomlj.{Toml, TomlArray, TomlInvalidTypeException, TomlParseResult, TomlTable}
 
-import java.io.{PrintStream, StringReader}
-import java.nio.file.Path
+import java.io.{IOException, PrintStream}
+import java.nio.file.{InvalidPathException, Path, Paths}
+import scala.collection.mutable
 
 object ManifestParser {
 
-  def parse(p: Path)(implicit out: PrintStream): Result[Manifest, ManifestError] = ???
+  def parse(p: Path)(implicit out: PrintStream): Result[Manifest, ManifestError] = {
+    val parser = try {
+      val parser = Toml.parse(p)
+      parser
+    } catch {
+      case _: IOException => return Err(ManifestError.IOError(p))
+    }
+
+    val errors = parser.errors
+    if(errors.size() > 0) {
+      var errorString = ""
+      errors.forEach(error => errorString = errorString + error.toString + ", ")
+      return Err(ManifestError.ManifestParseError(p, errorString))
+    }
+
+    val name = getRequiredStringProperty("package.name", parser, p) match {
+      case Ok(s) => s
+      case Err(e) => return Err(e)
+    }
+
+    val description = getRequiredStringProperty("package.description", parser, p) match {
```

Comment:
We should now be able to rewrite all of these using for-comprehensions like so:

```
for (
  description <- getRequiredStringProperty("package.description", parser, p)
  version <- getRequiredStringProperty("package.version", parser, p)
   // a lot more
) yield Ok(Manifest(name, description, semanticVersion, semanticFlixVersion, licenseOpt, authorsList, dependencyList))
```

Does that make sense?

Reply:
Thus we can avoid every single early return.